(auth): get tokens out of frontend, add `IAuthSession` interface with checked access

zachwolfe opened the pull request:
### Changes
- Auth tokens are no longer sent to the frontend. Instead, tokens are serialized and temporarily stored in a cookie registry on the backend, indexed by cryptographically secure ID. This ID is passed to the frontend and used to make an API call to the server, which then responds with the token as an HTTP-only cookie. Thus, the frontend never gets to see the token.
- All browser tabs with the same machine ID are now refreshed whenever the user logs _in_ (as opposed to #1562, which addressed the logging out case).
- A new `IAuthSession` interface is added which stores an `AuthToken` and a string that can be used to persist arbitrary session data into cookies. Most `IAuthProvider` methods now take this instead of a token.
- Due to the extreme flexibility provided by the `IAuthSession` interface (it allows getting and setting the token and session data at any time), extensive checks have been added and enabled in debug mode to aid with development. These take the form of the wrapper classes `CheckedAuthSession` and `CheckedAuthProvider`.
- `CheckedAuthSession` implements `IAuthSession` by wrapping another instance, but limiting access to `ReadOnly`, `WriteOnly`, or `ReadWrite` for each property (defaulting to no access allowed).
- `CheckedAuthProvider` implements `IAuthProvider` by wrapping another instance, but enforces `IAuthSession` access rules for each method using `CheckedAuthSession`. It also enforces some additional pre- and post-conditions.
- `IAuthProvider.SetHttpContext` has been refactored into `IAuthProvider.InitializeAsync`

zachwolfe commented:
I am going to expand the scope of this refactor so it can be more helpful for Clerk auth #1452

zachwolfe commented:
Now that I have tried these changes with the Clerk branch, I believe they are ready for review.

nielsbosma commented:
@zachwolfe why do we need the cookie concept? Why not use AppSessionStore?

Not sure I understand this PR fully. 

Also I think we should put everything that isn't a public API under Ivy.Core.

zachwolfe commented:
@nielsbosma I moved some things to `Ivy.Core` and moved all cookie registration functionality to `AppSessionStore`. The non-global `ICookieRegistry` stuff has been removed; the global timer will clean them up anyway.

Note that I had to add `AppSessionStore` as a service on the `Server` so that I could access it from `DefaultSidebarChrome` and `DefaultAuthApp`. Let me know if that is not OK with you.

zachwolfe commented:
@nielsbosma
> I'm thinking that instead of exposing AppSessionStore how about we put the logic for SetAuthCookies and SetAuthSessionData in IAuthService - IAuthService internally can access sessionStore and IClientProvider?

Yes, I like that idea.
